src/drawers/canvasBuffer.ts

```diff
@@ -0,0 +1,96 @@
+/**
+ * Copyright 2017-present Palantir Technologies
+ * @license MIT
+ */
+
+export class CanvasBuffer {
+    /**
+     * Resizes the canvas' internal pixel buffer to match the devicePixelRatio
+     */
+    public static sizePixels(
+            ctx: CanvasRenderingContext2D,
+            screenWidth: number,
+            screenHeight: number,
+            devicePixelRatio = 2,
+        ) {
+        const { canvas } = ctx;
+        canvas.width = screenWidth * devicePixelRatio;
+        canvas.height = screenHeight * devicePixelRatio;
+        canvas.style.width = `${screenWidth}px`;
+        canvas.style.height = `${screenHeight}px`;
+        ctx.setTransform(1, 0, 0, 1, 0, 0);
+        ctx.scale(devicePixelRatio, devicePixelRatio);
+    }
+
+    public canvas: HTMLCanvasElement;
+    public ctx: CanvasRenderingContext2D;
+    public pixelWidth: number;
+    public pixelHeight: number;
+
+    constructor(
+        public screenWidth: number,
+        public screenHeight: number,
+        public devicePixelRatio = 1,
```

Comment:
nit: this value can change (e.g. if the user drags a window from a non-retina into a retina display); we have no way of detecting when it changes so we should always pull from `window.devicePixelRatio` when rendering